### PR TITLE
typescriptify some init logic for functions and hosting

### LINKED ISSUE #4042: Convert to TS: src/init/features/index.js
The following file `src/init/features/index.js` should be converted to Typescript.

### LINKED ISSUE #4043: Convert to TS: src/init/features/hosting/index.js
The following file `src/init/features/hosting/index.js` should be converted to Typescript.

### PR DESCRIPTION
**Pull request by bkendall**
### Description

Typescripting some files.

Fixes #4042
Fixes #4043

### Scenarios Tested

init flows for `hosting` and `hosting:github` work - so does `functions`

### CODE REVIEW

File: src/init/features/index.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+export { doSetup as database } from "./database";
+export { doSetup as firestore } from "./firestore";
+export { doSetup as functions } from "./functions";
+export { doSetup as hosting } from "./hosting";
+export { doSetup as storage } from "./storage";
+export { doSetup as emulators } from "./emulators";
+// always runs, sets up .firebaserc
+export { doSetup as project } from "./project";
+export { doSetup as remoteconfig } from "./remoteconfig";
+export { initGitHub as hostingGithub } from "./hosting/github";
```
Comment (taeold) [inline src/init/features/index.ts:11]: 🔥

File: src/init/index.ts [hunk 2]
```diff
@@ -1,43 +1,50 @@
...
+      throw new FirebaseError(
+        `${clc.bold(nextFeature)} is not a valid feature. Must be one of ${availableFeatures}`
       );
     }
 
-    logger.info(clc.bold("\n" + clc.white("=== ") + _.capitalize(nextFeature) + " Setup"));
+    logger.info(clc.bold(`\n${clc.white("===")} ${capitalize(nextFeature)} Setup`));
 
-    await Promise.resolve(features[nextFeature](setup, config, options));
+    await Promise.resolve(featureFns.get(nextFeature)(setup, config, options));
```
Comment (taeold) [inline src/init/index.ts:47]: q: what does this extra `Promise.resolve` do 🤔

**Review by taeold — APPROVED** [review]

File: src/init/index.ts (hunk 2 above)
Comment (bkendall) [reply src/init/index.ts:47]: It ensures that whatever is being called _is_ a promise. If it's a promise, wrapping it one other promise is just a bit of overhead; if it's a sync function, it means that we await on it anyway. It's a safety check, more than anything. If any of the setup functions don't return a promise, this protected against that.

But now that it's all in typescript, and we _know_ the types of the functions, I can actually drop this extra promise! yay!

File: src/init/features/hosting/index.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      message: "Set up automatic builds and deploys with GitHub?",
+    },
+  ]);
+
+  setup.config.hosting = {
+    public: setup.hosting.public,
+    ignore: DEFAULT_IGNORES,
+  };
+
+  if (setup.hosting.spa) {
```
Comment (joehan) [inline src/init/features/hosting/index.ts:59]: Sounds very relaxing

**Review by joehan — APPROVED** [review]

**Pull request merged**